src/routes/tools/degree-days/_data.js

```diff
@@ -0,0 +1,220 @@
+// Node modules
+import { merge } from "d3-array";
+
+// Helpers
+import config from "~/helpers/api-config";
+import {
+  handleXHR,
+  fetchData,
+  transformResponse,
+  isLeapYear,
+} from "~/helpers/utilities";
+import { ENSEMBLES, OBSERVED, PRIORITY_10_MODELS } from "../_common/constants";
+import { buildEnvelope } from "../_common/helpers";
+
+const { apiEndpoint } = config.env.production;
+
+// Helper function to convert precipitation values from a rate (inches/day)
+// to total accumulation in a year
+const convertToAnnual = (values) => {
+  return values.map((d) => {
+    if (isLeapYear(+d.date.getFullYear())) {
+      return { ...d, value: d.value * 366 };
+    }
+    return { ...d, value: d.value * 365 };
+  });
+};
+
+/**
+ * The following 3 functions take the list of observed, models and ensemble
+ * raster series and add extra props. These props describe how the data
+ * should be fetched from the API or how the data is shown in the charts.
+ * List of extra props:
+ * slugs - list of raster series names in API
+ * mark - describes how the timeseries will be displayed in the chart (line/area)
+ * visible - controls whether line/area in chart is hidden/shown
+ *      when user clicks on corresponding legend key. Note: The stats component
+ *      also uses this prop to include/remove series from summary calculations
+ * @param {object} config
+ * @return {array}
+ */
+
+// The observed data is usually a single raster series, so only 1 slug
+const getObservedSeries = ({ climvarId }) => {
+  return OBSERVED.map((d) => {
+    const slugs = [`${climvarId}_day_${d.id}`];
+    return { ...d, slugs, mark: "line", visible: true };
+  });
+};
+
+// For each model, there are usually 2 raster series in the API,
+// the modeled historical (1950-2005) and modeled projections (2006-2099/2021)
+const getModelSeries = ({ climvarId, scenarioId, modelIds, indicatorId }) => {
+  return PRIORITY_10_MODELS.filter((d) => modelIds.includes(d.id)).map((d) => {
+    const slugs = [
+      // `${climvarId}_day_${d.id}_historical`,
+      `${climvarId}_day_${d.id}_${scenarioId}`,
+    ];
+    return { ...d, slugs, mark: "line", visible: true, indicatorId };
+  });
+};
+
+// The ensemble has to be assembled from the max and min of 10/all models
+// Similar to the models, the models-max and models-min are 2 raster series each
+const getEnsembleSeries = ({ climvarId, scenarioId }) => {
+  return ENSEMBLES.filter((d) => d.id === `${scenarioId}_range`).map((d) => {
+    const slugs = [
+      `${climvarId}_year_models-min_historical`,
```

Comment:
@mukhtyar I tried changing these to one listed in the google doc but it only returned a list of rasters, not actual data. I couldn't find more info the API docs, do you know what the correct API call would be? For example:

```
https://api.cal-adapt.org/api/series/cdd_year_ens32avg_historical/?imperial=true&g=Point(-121.4688%2038.5938)
```

Reply:
Oh actually nevermind, I see that appending `/events/` to the URL path returns data.

Reply:
Yes. Although I just realized that these are pre-calculated CDD/HDD values with reference temperature set to 65. So if the user selects a different reference temp, these numbers would not update. In that case, we can just stick with the observed & models and not add the ensembles. Let's confirm this at our meeting today.

Reply:
Got it, good catch

Reply:
If you'd like to add ensembles for CDD and HDD, you can compute them dynamically from the various tasmax 32 GCM ensembles with varying thresholds and frequencies like:

https://api.cal-adapt.org/api/series/tasmax_day_ens32avg_rcp85/cdd/?freq=A&thresh=65&g=Point(-121.4688+38.59)

Reply:
Thanks @bkg that's helpful to know. 

In a somewhat related question, for the Cal-Adapt API is there a way to request specific months for the `freq` param? For example something like `freq=[5,6,7]` for grabbing only the months of June, July, and August. 

Currently in the dashboard we're filtering and then re-aggregating clientside; but from my basic stats background I don't think it's a good idea to do an average of averages, in this case averaging the filtered months from the API response which I assume themselves are already averages.

Reply:
The `freq` param is for temporal aggregation and it supports any of the [Pandas offsets](http://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#dateoffset-objects). To retrieve an average of those particular months, you can request a quarterly frequency starting in May:

https://api.cal-adapt.org/api/series/tasmax_day_ens32avg_rcp85/cdd/?freq=QS-MAY&thresh=65&g=Point(-121.4688+38.59)

You can fine tune the number of months as well, so `2MS` would give you a bimonthly frequency.

Reply:
The issue is that the current (v2) app allows users to select arbitrary months, so a user could select January and then June (which wouldn't make any sense). We would need to redesign that part of the UI to restrict it to quarterly periods, which I'm not sure our users would object to or not.

How does the API aggregate CCD & HDD? Is it summing them based on `freq`?

Reply:
Ah, okay you could do any number of sequential months with the API, not limited to quarters, but you couldn't group non-sequential months with the resampler.

The default aggregation is `?stat=mean` if not specified, but more summary stats are available like min, max. The [docs](https://berkeley-gif.github.io/caladapt-docs/series.html#time-series) have some more detail.

Reply:
I think it makes more sense to limit it to sequential months, but we would need to rethink the UI for the frequency selection which I'm not sure we have time for right now. Definitely worth brainstorming on. 

For the immediate time being if we continue with our current approach for monthly frequency by passing`freq=M`, and then filter and aggregate clientside for only the months the user has selected, would it be correct to also pass `stat=sum` and then average the results when aggregating client side?

Reply:
Context: I'm trying to avoid computing an average of averages ([here's a write up on why that's a bad practice](https://convalytics.com/statistics/taking-an-average-of-averages-why-its-wrong/)). Because months differ in the number of days they have, we probably shouldn't be averaging the average number of CDD/HDD for each month by the number of months the user has selected. We should instead be averaging by the total number of CDD/HDD by the total number of days for all selected months.

Reply:
The `stat` param is related to spatial aggregation across grid cells and not temporal aggregation. So in case of temperature related indicators, if you want temperature for a polygon, e.g county, you get a temperature value for the county by taking the `mean` of temperatures across all intersecting grid cells. In case of precipitation amount for a county, it makes more sense to `sum` the amount of rain in each grid cell. On the other hand if you want rate of precipitation, then you might use `mean`. 

In the CDD/HDD calculations, for a grid cell, the values returned for each month/year (indicated by the `freq` param) is the total number of CDD/HDD for that month/year. This also applies to values returned for a county. So I think averaging the totals over `n` months in the tool is ok. 

When running the calculations for a county in the API, I think the spatial aggregation happens first. So we have a new daily timeseries of tasmax and tasmin for the county, and the CDD/HDD calculation & any temporal aggregation happens next on this new timeseries. (not sure if this is the sequence of events or if CDD/HDD values are calculated for each grid cell and then aggregated over the county; @bkg can confirm)

Reply:
Sorry, I misspoke about the `stat` param for the CDD and HDD endpoints which differs from events. For CDD/HDD, `stat` controls spatial, not temporal, aggregation. You'll want to keep using mean for spatial aggregation, which is the default, so just passing `freq=M` should be enough.

Reply:
Okay, got it. Thanks for clarifying @mukhtyar & @bkg

Reply:
> 
> 
> If you'd like to add ensembles for CDD and HDD, you can compute them dynamically from the various tasmax 32 GCM ensembles with varying thresholds and frequencies like:
> 
> https://api.cal-adapt.org/api/series/tasmax_day_ens32avg_rcp85/cdd/?freq=A&thresh=65&g=Point(-121.4688+38.59)

@bkg weren't there issues with calculating an envelope from the daily ensemble max and min `tasmax_day_ens32max_rcp85` and `tasmax_day_ens32min_rcp85`? It returned the most extreme values? Or was that only an issue for the Extreme Heat Day calculations? 

To show the ranges for Extreme Heat Days in the Location Climate Change Snapshot tool we ended up creating new timeseries `exheat_30y_ens32max_rcp85` and `exheat_30y_ens32min_rcp85`

Reply:
That's correct. Generally it's an issue for the more complicated climate indices such as extreme heat, KBDI, but not the more straightforward ones like CDD and HDD which only rely on number of degrees above/below a static threshold instead of a percentile like extreme heat. 

If it's problematic, then it will be evident from the ensemble min/max as you mentioned, but I think it will be okay for these indices. Plus, it allows using a dynamic threshold which the pre-generated grids do not.

Reply:
Got it, thanks! We will try it out.